File: src/bmsx_cpp/emulator/devices/dma_controller.cpp
#include "dma_controller.h"

#include "../io.h"
#include "../memory_map.h"

#include <algorithm>
#include <utility>

namespace bmsx {

DmaController::DmaController(
	Memory& memory,
	std::function<void(uint32_t)> raiseIrq,
	std::function<void(uint32_t src, size_t length)> sealVdpFifoDma
)
	: m_memory(memory)
	, m_raiseIrq(std::move(raiseIrq))
	, m_sealVdpFifoDma(std::move(sealVdpFifoDma)) {}

bool DmaController::hasPendingVdpSubmit() const {
	for (const auto& state : m_channels) {
		if (state.hasActive && state.active.kind == DmaJob::Kind::Io && state.active.dst == IO_VDP_FIFO) {
			return true;
		}
		for (const auto& job : state.queue) {
			if (job.kind == DmaJob::Kind::Io && job.dst == IO_VDP_FIFO) {
				return true;
			}
		}
	}
	return false;
}

void DmaController::setChannelBudgets(uint32_t isoBytesPerTick, uint32_t bulkBytesPerTick) {
	m_channels[static_cast<int>(Channel::Iso)].budget = isoBytesPerTick;
	m_channels[static_cast<int>(Channel::Bulk)].budget = bulkBytesPerTick;
}

void DmaController::enqueueImageCopy(const Memory::ImageWritePlan& plan, std::vector<uint8_t>&& pixels, std::function<void(bool error, bool clipped)> onComplete) {
	DmaJob job;
	job.kind = DmaJob::Kind::Image;
	job.channel = Channel::Bulk;
	job.plan = plan;
	job.pixels = std::move(pixels);
	job.row = 0;
	job.rowOffset = 0;
	job.vramTarget = m_memory.isVramRange(plan.baseAddr, plan.writeLen > 0 ? plan.writeLen : 1);
	job.written = 0;
	job.clipped = plan.clipped;
	job.error = false;
	job.onComplete = std::move(onComplete);
	m_channels[static_cast<int>(Channel::Bulk)].queue.push_back(std::move(job));
}

void DmaController::tick() {
	bool ioWrittenDirty = false;
	bool imgWrittenDirty = false;
	tickChannel(Channel::Iso, ioWrittenDirty, imgWrittenDirty);
	tickChannel(Channel::Bulk, ioWrittenDirty, imgWrittenDirty);
	if (ioWrittenDirty) {
		m_memory.writeValue(IO_DMA_WRITTEN, valueNumber(static_cast<double>(m_ioWrittenValue)));
	}
	if (imgWrittenDirty) {
		m_memory.writeValue(IO_IMG_WRITTEN, valueNumber(static_cast<double>(m_imgWrittenValue)));
	}
}

void DmaController::reset() {
	for (int i = 0; i < 2; i += 1) {
		m_channels[i].queue.clear();
		m_channels[i].hasActive = false;
	}
	m_ioWrittenValue = 0;
	m_imgWrittenValue = 0;
	m_buffer.clear();
	m_memory.writeValue(IO_DMA_SRC, valueNumber(0.0));
	m_memory.writeValue(IO_DMA_DST, valueNumber(0.0));
	m_memory.writeValue(IO_DMA_LEN, valueNumber(0.0));
	m_memory.writeValue(IO_DMA_CTRL, valueNumber(0.0));
	m_memory.writeValue(IO_DMA_STATUS, valueNumber(0.0));
	m_memory.writeValue(IO_DMA_WRITTEN, valueNumber(0.0));
}

void DmaController::tickChannel(Channel channel, bool& ioWrittenDirty, bool& imgWrittenDirty) {
	auto& state = m_channels[static_cast<int>(channel)];
	uint32_t budget = state.budget;
	while (budget > 0) {
		if (!state.hasActive) {
			if (state.queue.empty()) {
				return;
			}
			state.active = std::move(state.queue.front());
			state.queue.pop_front();
			state.hasActive = true;
		}
		auto& job = state.active;
		const uint32_t written = processJob(job, budget);
		budget -= written;
		if (job.kind == DmaJob::Kind::Io) {
			m_ioWrittenValue = job.written;
			ioWrittenDirty = true;
		}
		if (job.kind == DmaJob::Kind::Image) {
			m_imgWrittenValue = job.written;
			imgWrittenDirty = true;
		}
		if (isJobComplete(job)) {
			finishJob(job);
			state.hasActive = false;
			continue;
		}
		if (written == 0) {
			return;
		}
	}
}

uint32_t DmaController::processJob(DmaJob& job, uint32_t budget) {
	if (job.error) {
		return 0;
	}
	if (job.kind == DmaJob::Kind::Io) {
		uint32_t chunk = job.remaining > budget ? budget : job.remaining;
		if (chunk == 0) {
			return 0;
		}
		if (job.dst == IO_VDP_FIFO) {
			job.remaining -= chunk;
			job.written += chunk;
			return chunk;
		}
		if (m_memory.isVramRange(job.dst, 1)) {
			chunk &= ~3u;
			if (chunk == 0) {
				return 0;
			}
		}
		try {
			if (m_buffer.size() < chunk) {
				m_buffer.resize(chunk);
			}
			m_memory.readBytes(job.src, m_buffer.data(), chunk);
			m_memory.writeBytes(job.dst, m_buffer.data(), chunk);
		} catch (...) {
			job.error = true;
			return 0;
		}
		job.src += chunk;
		job.dst += chunk;
		job.remaining -= chunk;
		job.written += chunk;
		return chunk;
	}
	return processImageJob(job, budget);
}

uint32_t DmaController::processImageJob(DmaJob& job, uint32_t budget) {
	uint32_t remaining = budget;
	while (remaining > 0 && job.row < job.plan.writeHeight) {
		const uint32_t rowRemaining = job.plan.writeStride - job.rowOffset;
		uint32_t toCopy = remaining < rowRemaining ? remaining : rowRemaining;
		if (job.vramTarget) {
			toCopy &= ~3u;
			if (toCopy == 0) {
				return budget - remaining;
			}
		}
		const size_t srcOffset = static_cast<size_t>(job.row) * job.plan.sourceStride + job.rowOffset;
		const uint32_t dstAddr = job.plan.baseAddr + job.row * job.plan.targetStride + job.rowOffset;
		try {
			m_memory.writeBytes(dstAddr, job.pixels.data() + srcOffset, toCopy);
		} catch (...) {
			job.error = true;
			return budget - remaining;
		}
		remaining -= toCopy;
		job.rowOffset += toCopy;
		job.written += toCopy;
		if (job.rowOffset >= job.plan.writeStride) {
			job.row += 1;
			job.rowOffset = 0;
		}
	}
	return budget - remaining;
}

bool DmaController::isJobComplete(const DmaJob& job) const {
	if (job.error) {
		return true;
	}
	if (job.kind == DmaJob::Kind::Io) {
		return job.remaining == 0;
	}
	return job.row >= job.plan.writeHeight;
}

void DmaController::finishJob(DmaJob& job) {
	if (job.kind == DmaJob::Kind::Io) {
		finishIoJob(job);
		return;
	}
	if (job.onComplete) {
		job.onComplete(job.error, job.clipped);
	}
}

void DmaController::tryStartIo() {
	const uint32_t ctrlValue = static_cast<uint32_t>(asNumber(m_memory.readValue(IO_DMA_CTRL)));
	if ((ctrlValue & DMA_CTRL_START) == 0) {
		return;
	}
	const uint32_t ctrl = ctrlValue;
	const uint32_t src = static_cast<uint32_t>(asNumber(m_memory.readValue(IO_DMA_SRC)));
	const uint32_t dst = static_cast<uint32_t>(asNumber(m_memory.readValue(IO_DMA_DST)));
	const uint32_t len = static_cast<uint32_t>(asNumber(m_memory.readValue(IO_DMA_LEN)));
	const bool strict = (ctrl & DMA_CTRL_STRICT) != 0;
	m_memory.writeValue(IO_DMA_CTRL, valueNumber(static_cast<double>(ctrl & ~DMA_CTRL_START)));
	if (dst == IO_VDP_FIFO && hasPendingVdpSubmit()) {
		finishIoRejected();
		return;
	}
	m_memory.writeValue(IO_DMA_WRITTEN, valueNumber(0.0));

	const uint32_t maxWritable = resolveMaxWritable(dst);
	if (maxWritable == 0) {
		finishIoError(false);
		return;
	}
	uint32_t transferLen = len;
	bool clipped = false;
	if (transferLen > maxWritable) {
		clipped = true;
		if (strict) {
			finishIoError(true);
			return;
		}
		transferLen = maxWritable;
	}
	const uint32_t status = DMA_STATUS_BUSY | (clipped ? DMA_STATUS_CLIPPED : 0);
	m_memory.writeValue(IO_DMA_STATUS, valueNumber(static_cast<double>(status)));
	if (transferLen == 0) {
		finishIoSuccess(clipped);
		return;
	}
	DmaJob job;
	job.kind = DmaJob::Kind::Io;
	job.channel = Channel::Bulk;
	job.src = src;
	job.dst = dst;
	job.remaining = transferLen;
	job.written = 0;
	job.clipped = clipped;
	job.strict = strict;
	job.error = false;
	m_ioWrittenValue = 0;
	m_channels[static_cast<int>(Channel::Bulk)].queue.push_back(std::move(job));
}

void DmaController::finishIoJob(DmaJob& job) {
	if (job.error) {
		finishIoError(job.clipped);
		return;
	}
	if (job.dst == IO_VDP_FIFO) {
		try {
			m_sealVdpFifoDma(job.src, job.written);
		} catch (...) {
			finishIoError(job.clipped);
			return;
		}
	}
	finishIoSuccess(job.clipped);
}

void DmaController::finishIoSuccess(bool clipped) {
	uint32_t status = DMA_STATUS_DONE;
	if (clipped) {
		status |= DMA_STATUS_CLIPPED;
	}
	m_memory.writeValue(IO_DMA_STATUS, valueNumber(static_cast<double>(status)));
	m_raiseIrq(IRQ_DMA_DONE);
}

void DmaController::finishIoError(bool clipped) {
	uint32_t status = DMA_STATUS_DONE | DMA_STATUS_ERROR;
	if (clipped) {
		status |= DMA_STATUS_CLIPPED;
	}
	m_memory.writeValue(IO_DMA_STATUS, valueNumber(static_cast<double>(status)));
	m_raiseIrq(IRQ_DMA_ERROR);
}

void DmaController::finishIoRejected() {
	m_memory.writeValue(IO_DMA_WRITTEN, valueNumber(0.0));
	m_memory.writeValue(IO_DMA_STATUS, valueNumber(static_cast<double>(DMA_STATUS_REJECTED)));
}

uint32_t DmaController::resolveMaxWritable(uint32_t dst) const {
	if (dst == IO_VDP_FIFO) {
		return VDP_STREAM_BUFFER_SIZE;
	}
	if (dst >= VRAM_SYSTEM_ATLAS_BASE && dst < VRAM_SYSTEM_ATLAS_BASE + VRAM_SYSTEM_ATLAS_SIZE) {
		return (VRAM_SYSTEM_ATLAS_BASE + VRAM_SYSTEM_ATLAS_SIZE) - dst;
	}
	if (dst >= VRAM_SKYBOX_BASE && dst < VRAM_SKYBOX_BASE + VRAM_SKYBOX_SIZE) {
		return (VRAM_SKYBOX_BASE + VRAM_SKYBOX_SIZE) - dst;
	}
	if (dst >= VRAM_PRIMARY_ATLAS_BASE && dst < VRAM_PRIMARY_ATLAS_BASE + VRAM_PRIMARY_ATLAS_SIZE) {
		return (VRAM_PRIMARY_ATLAS_BASE + VRAM_PRIMARY_ATLAS_SIZE) - dst;
	}
	if (dst >= VRAM_SECONDARY_ATLAS_BASE && dst < VRAM_SECONDARY_ATLAS_BASE + VRAM_SECONDARY_ATLAS_SIZE) {
		return (VRAM_SECONDARY_ATLAS_BASE + VRAM_SECONDARY_ATLAS_SIZE) - dst;
	}
	if (dst >= VRAM_FRAMEBUFFER_BASE && dst < VRAM_FRAMEBUFFER_BASE + VRAM_FRAMEBUFFER_SIZE) {
		return (VRAM_FRAMEBUFFER_BASE + VRAM_FRAMEBUFFER_SIZE) - dst;
	}
	if (dst >= VRAM_STAGING_BASE && dst < VRAM_STAGING_BASE + VRAM_STAGING_SIZE) {
		return (VRAM_STAGING_BASE + VRAM_STAGING_SIZE) - dst;
	}
	if (dst >= OVERLAY_ROM_BASE && dst < OVERLAY_ROM_BASE + OVERLAY_ROM_SIZE) {
		return (OVERLAY_ROM_BASE + OVERLAY_ROM_SIZE) - dst;
	}
	if (dst >= RAM_BASE && dst < RAM_USED_END) {
		return RAM_USED_END - dst;
	}
	return 0;
}

} // namespace bmsx

File: src/bmsx/emulator/devices/dma_controller.ts
import {
	DMA_CTRL_START,
	DMA_CTRL_STRICT,
	DMA_STATUS_BUSY,
	DMA_STATUS_CLIPPED,
	DMA_STATUS_DONE,
	DMA_STATUS_ERROR,
	DMA_STATUS_REJECTED,
	IO_DMA_CTRL,
	IO_DMA_DST,
	IO_DMA_LEN,
	IO_DMA_SRC,
	IO_DMA_STATUS,
	IO_DMA_WRITTEN,
	IO_IMG_WRITTEN,
	IO_VDP_FIFO,
	IRQ_DMA_DONE,
	IRQ_DMA_ERROR,
} from '../io';
import {
	OVERLAY_ROM_BASE,
	OVERLAY_ROM_SIZE,
	RAM_BASE,
	RAM_USED_END,
	VDP_STREAM_BUFFER_SIZE,
	VRAM_SYSTEM_ATLAS_BASE,
	VRAM_SYSTEM_ATLAS_SIZE,
	VRAM_PRIMARY_ATLAS_BASE,
	VRAM_PRIMARY_ATLAS_SIZE,
	VRAM_FRAMEBUFFER_BASE,
	VRAM_FRAMEBUFFER_SIZE,
	VRAM_SKYBOX_BASE,
	VRAM_SKYBOX_SIZE,
	VRAM_SECONDARY_ATLAS_BASE,
	VRAM_SECONDARY_ATLAS_SIZE,
	VRAM_STAGING_BASE,
	VRAM_STAGING_SIZE,
} from '../memory_map';
import type { ImageWritePlan } from '../memory';
import { Memory } from '../memory';

type DmaChannelId = 0 | 1;
const DMA_CH_ISO: DmaChannelId = 0;
const DMA_CH_BULK: DmaChannelId = 1;

type DmaJobBase = {
	kind: 'io' | 'image';
	channel: DmaChannelId;
	written: number;
	clipped: boolean;
	error: boolean;
	fault: unknown;
};

type DmaIoJob = DmaJobBase & {
	kind: 'io';
	src: number;
	dst: number;
	remaining: number;
	strict: boolean;
};

type DmaImageJob = DmaJobBase & {
	kind: 'image';
	plan: ImageWritePlan;
	pixels: Uint8Array;
	row: number;
	rowOffset: number;
	vramTarget: boolean;
	onComplete: (result: { error: boolean; clipped: boolean; fault: unknown }) => void;
};

type DmaJob = DmaIoJob | DmaImageJob;

type DmaChannelState = {
	budget: number;
	queue: DmaJob[];
	active: DmaJob | null;
};

export class DmaController {
	private readonly channels: [DmaChannelState, DmaChannelState] = [
		{ budget: 0, queue: [], active: null },
		{ budget: 0, queue: [], active: null },
	];
	private ioWrittenValue = 0;
	private ioWrittenDirty = false;
	private imgWrittenValue = 0;
	private imgWrittenDirty = false;

	public constructor(
		private readonly memory: Memory,
		private readonly raiseIrq: (mask: number) => void,
		private readonly sealVdpFifoDma: (src: number, length: number) => void,
	) {}

	public hasPendingVdpSubmit(): boolean {
		for (let channelIndex = 0; channelIndex < this.channels.length; channelIndex += 1) {
			const state = this.channels[channelIndex]!;
			if (state.active !== null && state.active.kind === 'io' && state.active.dst === IO_VDP_FIFO) {
				return true;
			}
			for (let queueIndex = 0; queueIndex < state.queue.length; queueIndex += 1) {
				const job = state.queue[queueIndex]!;
				if (job.kind === 'io' && job.dst === IO_VDP_FIFO) {
					return true;
				}
			}
		}
		return false;
	}

	public setChannelBudgets(params: { iso: number; bulk: number }): void {
		this.channels[DMA_CH_ISO].budget = params.iso;
		this.channels[DMA_CH_BULK].budget = params.bulk;
	}

	public reset(): void {
		this.channels[DMA_CH_ISO].queue.length = 0;
		this.channels[DMA_CH_ISO].active = null;
		this.channels[DMA_CH_BULK].queue.length = 0;
		this.channels[DMA_CH_BULK].active = null;
		this.ioWrittenValue = 0;
		this.ioWrittenDirty = false;
		this.imgWrittenValue = 0;
		this.imgWrittenDirty = false;
		this.memory.writeValue(IO_DMA_SRC, 0);
		this.memory.writeValue(IO_DMA_DST, 0);
		this.memory.writeValue(IO_DMA_LEN, 0);
		this.memory.writeValue(IO_DMA_CTRL, 0);
		this.memory.writeValue(IO_DMA_STATUS, 0);
		this.memory.writeValue(IO_DMA_WRITTEN, 0);
	}

	public enqueueImageCopy(plan: ImageWritePlan, pixels: Uint8Array, onComplete: (result: { error: boolean; clipped: boolean; fault: unknown }) => void): void {
		const vramTarget = this.memory.isVramRange(plan.baseAddr, plan.writeSize > 0 ? plan.writeSize : 1);
		const job: DmaImageJob = {
			kind: 'image',
			channel: DMA_CH_BULK,
			plan,
			pixels,
			row: 0,
			rowOffset: 0,
			vramTarget,
			written: 0,
			clipped: plan.clipped,
			error: false,
			fault: null,
			onComplete,
		};
		this.channels[DMA_CH_BULK].queue.push(job);
	}

	public tick(): void {
		this.tickChannel(DMA_CH_ISO);
		this.tickChannel(DMA_CH_BULK);
		if (this.ioWrittenDirty) {
			this.memory.writeValue(IO_DMA_WRITTEN, this.ioWrittenValue);
			this.ioWrittenDirty = false;
		}
		if (this.imgWrittenDirty) {
			this.memory.writeValue(IO_IMG_WRITTEN, this.imgWrittenValue);
			this.imgWrittenDirty = false;
		}
	}

	public tryStartIo(): void {
		const ctrlValue = this.memory.readValue(IO_DMA_CTRL) as number;
		if ((ctrlValue & DMA_CTRL_START) === 0) {
			return;
		}
		const ctrl = ctrlValue >>> 0;
		const src = (this.memory.readValue(IO_DMA_SRC) as number) >>> 0;
		const dst = (this.memory.readValue(IO_DMA_DST) as number) >>> 0;
		const len = (this.memory.readValue(IO_DMA_LEN) as number) >>> 0;
		const vdpSubmit = dst === IO_VDP_FIFO;
		const strict = (ctrl & DMA_CTRL_STRICT) !== 0;
		this.memory.writeValue(IO_DMA_CTRL, ctrl & ~DMA_CTRL_START);
		if (vdpSubmit && this.hasPendingVdpSubmit()) {
			this.finishIoRejected();
			return;
		}
		this.memory.writeValue(IO_DMA_WRITTEN, 0);
		if (vdpSubmit && (len & 3) !== 0) {
			this.finishIoError(false);
			return;
		}
		const maxWritable = this.resolveMaxWritable(dst);
		if (maxWritable <= 0) {
			this.finishIoError(false);
			return;
		}
		let transferLen = len;
		let clipped = false;
		if (transferLen > maxWritable) {
			clipped = true;
			if (strict || vdpSubmit) {
				this.finishIoError(true);
				return;
			}
			transferLen = maxWritable;
		}
		const status = DMA_STATUS_BUSY | (clipped ? DMA_STATUS_CLIPPED : 0);
		this.memory.writeValue(IO_DMA_STATUS, status);
		if (transferLen === 0) {
			this.finishIoSuccess(clipped);
			return;
		}
		const job: DmaIoJob = {
			kind: 'io',
			channel: DMA_CH_BULK,
			src,
			dst,
			remaining: transferLen,
			written: 0,
			clipped,
			strict,
			error: false,
			fault: null,
		};
		this.ioWrittenValue = 0;
		this.ioWrittenDirty = true;
		this.channels[DMA_CH_BULK].queue.push(job);
	}

	private tickChannel(channel: DmaChannelId): void {
		const state = this.channels[channel];
		let budget = state.budget;
		while (budget > 0) {
			if (!state.active) {
				const next = state.queue.shift();
				if (!next) {
					return;
				}
				state.active = next;
			}
			const job = state.active;
			const written = this.processJob(job, budget);
			budget -= written;
			if (job.kind === 'io') {
				this.ioWrittenValue = job.written;
				this.ioWrittenDirty = true;
			}
			if (job.kind === 'image') {
				this.imgWrittenValue = job.written;
				this.imgWrittenDirty = true;
			}
			if (this.isJobComplete(job)) {
				this.finishJob(job);
				state.active = null;
				continue;
			}
			if (written === 0) {
				return;
			}
		}
	}

	private processJob(job: DmaJob, budget: number): number {
		if (job.error) {
			return 0;
		}
		if (job.kind === 'io') {
			let chunk = job.remaining > budget ? budget : job.remaining;
			if (chunk === 0) {
				return 0;
			}
			if (job.dst === IO_VDP_FIFO) {
				job.remaining -= chunk;
				job.written += chunk;
				return chunk;
			}
			if (this.memory.isVramRange(job.dst, 1)) {
				chunk &= ~3;
				if (chunk === 0) {
					return 0;
				}
			}
			try {
				const bytes = this.memory.readBytes(job.src, chunk);
				this.memory.writeBytes(job.dst, bytes);
			} catch (error) {
				job.error = true;
				job.fault = error;
				return 0;
			}
			job.src += chunk;
			job.dst += chunk;
			job.remaining -= chunk;
			job.written += chunk;
			return chunk;
		}
		return this.processImageJob(job, budget);
	}

	private processImageJob(job: DmaImageJob, budget: number): number {
		let remaining = budget;
		while (remaining > 0 && job.row < job.plan.writeHeight) {
			const rowRemaining = job.plan.writeStride - job.rowOffset;
			let toCopy = remaining < rowRemaining ? remaining : rowRemaining;
			if (job.vramTarget) {
				toCopy &= ~3;
				if (toCopy === 0) {
					return budget - remaining;
				}
			}
			const srcOffset = job.row * job.plan.sourceStride + job.rowOffset;
			const dstAddr = job.plan.baseAddr + (job.row * job.plan.targetStride) + job.rowOffset;
			try {
				this.memory.writeBytesFrom(job.pixels, srcOffset, dstAddr, toCopy);
			} catch (error) {
				job.error = true;
				job.fault = error;
				return budget - remaining;
			}
			remaining -= toCopy;
			job.rowOffset += toCopy;
			job.written += toCopy;
			if (job.rowOffset >= job.plan.writeStride) {
				job.row += 1;
				job.rowOffset = 0;
			}
		}
		return budget - remaining;
	}

	private isJobComplete(job: DmaJob): boolean {
		if (job.error) {
			return true;
		}
		if (job.kind === 'io') {
			return job.remaining === 0;
		}
		return job.row >= job.plan.writeHeight;
	}

	private finishJob(job: DmaJob): void {
		if (job.kind === 'io') {
			this.finishIoJob(job);
			return;
		}
		job.onComplete({ error: job.error, clipped: job.clipped, fault: job.fault });
	}

	private resolveMaxWritable(dst: number): number {
		if (dst === IO_VDP_FIFO) {
			return VDP_STREAM_BUFFER_SIZE;
		}
		if (dst >= VRAM_SYSTEM_ATLAS_BASE && dst < VRAM_SYSTEM_ATLAS_BASE + VRAM_SYSTEM_ATLAS_SIZE) {
			return (VRAM_SYSTEM_ATLAS_BASE + VRAM_SYSTEM_ATLAS_SIZE) - dst;
		}
		if (dst >= VRAM_PRIMARY_ATLAS_BASE && dst < VRAM_PRIMARY_ATLAS_BASE + VRAM_PRIMARY_ATLAS_SIZE) {
			return (VRAM_PRIMARY_ATLAS_BASE + VRAM_PRIMARY_ATLAS_SIZE) - dst;
		}
		if (dst >= VRAM_SECONDARY_ATLAS_BASE && dst < VRAM_SECONDARY_ATLAS_BASE + VRAM_SECONDARY_ATLAS_SIZE) {
			return (VRAM_SECONDARY_ATLAS_BASE + VRAM_SECONDARY_ATLAS_SIZE) - dst;
		}
		if (dst >= VRAM_FRAMEBUFFER_BASE && dst < VRAM_FRAMEBUFFER_BASE + VRAM_FRAMEBUFFER_SIZE) {
			return (VRAM_FRAMEBUFFER_BASE + VRAM_FRAMEBUFFER_SIZE) - dst;
		}
		if (dst >= VRAM_STAGING_BASE && dst < VRAM_STAGING_BASE + VRAM_STAGING_SIZE) {
			return (VRAM_STAGING_BASE + VRAM_STAGING_SIZE) - dst;
		}
		if (dst >= VRAM_SKYBOX_BASE && dst < VRAM_SKYBOX_BASE + VRAM_SKYBOX_SIZE) {
			return (VRAM_SKYBOX_BASE + VRAM_SKYBOX_SIZE) - dst;
		}
		if (dst >= OVERLAY_ROM_BASE && dst < OVERLAY_ROM_BASE + OVERLAY_ROM_SIZE) {
			return (OVERLAY_ROM_BASE + OVERLAY_ROM_SIZE) - dst;
		}
		if (dst >= RAM_BASE && dst < RAM_USED_END) {
			return RAM_USED_END - dst;
		}
		return 0;
	}

	private finishIoJob(job: DmaIoJob): void {
		if (job.error) {
			this.finishIoError(job.clipped);
			return;
		}
		if (job.dst === IO_VDP_FIFO) {
			try {
				this.sealVdpFifoDma(job.src, job.written);
			} catch {
				this.finishIoError(job.clipped);
				return;
			}
		}
		this.finishIoSuccess(job.clipped);
	}

	private finishIoSuccess(clipped: boolean): void {
		let status = DMA_STATUS_DONE;
		if (clipped) {
			status |= DMA_STATUS_CLIPPED;
		}
		this.memory.writeValue(IO_DMA_STATUS, status);
		this.raiseIrq(IRQ_DMA_DONE);
	}

	private finishIoError(clipped: boolean): void {
		let status = DMA_STATUS_DONE | DMA_STATUS_ERROR;
		if (clipped) {
			status |= DMA_STATUS_CLIPPED;
		}
		this.memory.writeValue(IO_DMA_STATUS, status);
		this.raiseIrq(IRQ_DMA_ERROR);
	}

	private finishIoRejected(): void {
		this.memory.writeValue(IO_DMA_WRITTEN, 0);
		this.memory.writeValue(IO_DMA_STATUS, DMA_STATUS_REJECTED);
	}
}
